### PR TITLE
Add buy ticket logic and tests to it.

### PR DESCRIPTION
**Pull request by bodikqlar**
Add prawn gem for pdf download.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -39,10 +39,15 @@ gem 'capybara'
 gem 'capybara-webkit'
 #https://github.com/DatabaseCleaner/database_cleaner
 gem 'database_cleaner'
-
+#https://github.com/apotonick/cells
```
Comment (MaximSt) [inline Gemfile]: https://github.com/apotonick/cells/commit/943c2b102078a933ac93f0d70f32e62044ab11c9

File: Gemfile [hunk 2]
```diff
@@ -39,10 +39,15 @@ gem 'capybara'
...
+gem 'cells', '~> 4.0.0.beta6', git: 'https://github.com/apotonick/cells.git'
```
Comment (maksymst) [inline Gemfile]: https://github.com/apotonick/cells/commit/943c2b102078a933ac93f0d70f32e62044ab11c9

File: app/cells/cart_cell.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  include Cell::Haml
+  property :cart_items
+
+  def show
+    render
+  end
+
+  private
+
+  def quantity
```
Comment (maksymst) [inline app/cells/cart_cell.rb]: move to `cart` model

File: app/cells/cart_cell.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    cart_items.inject(0){|result, current| result + current.quantity}
+  end
+
+  def price
```
Comment (maksymst) [inline app/cells/cart_cell.rb]: move to `cart` model

File: app/cells/cart_cell.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
+class CartCell < Cell::ViewModel
+  include ActionView::Helpers::FormHelper
+  include ActionView::Helpers::NumberHelper
+  include Cell::Haml
+  property :cart_items
```
Comment (maksymst) [inline app/cells/cart_cell.rb]: replace with `cart`

File: app/controllers/movies_controller.rb [hunk 6]
```diff
@@ -23,8 +23,9 @@ def new
   end
 
   def create
-    @movie = current_user.movies.create(movie_params)
+    @movie = current_user.movies.build(movie_params)
     if @movie.save
+      Ticket.create_tickets_for_movie(@movie)
```
Comment (maksymst) [inline app/controllers/movies_controller.rb]: must be a callback

File: app/controllers/movies_controller.rb [hunk 7]
```diff
@@ -50,6 +51,16 @@ def update
     end
   end
 
+  def add_to_the_cart
+    if current_user.cart
```
Comment (maksymst) [inline app/controllers/movies_controller.rb]: move conditional login to model

File: app/controllers/movies_controller.rb [hunk 8]
```diff
@@ -68,5 +79,11 @@ def get_categories
     @categories = Category.pluck(:name, :id)
   end
 
-
+  def add_cart_items_to_cart
```
Comment (maksymst) [inline app/controllers/movies_controller.rb]: move conditional login to model

File: app/helpers/application_helper.rb [hunk 9]
```diff
@@ -12,4 +12,7 @@ def errors_output
       end.join().html_safe
     end
   end
+  def calculate_sum(quantity, price)
+    number_to_currency(quantity*price)
```
Comment (maksymst) [inline app/helpers/application_helper.rb]: move login to model

File: app/models/ticket.rb [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+  validates_numericality_of :price, greater_than: 0
+  validates_numericality_of :place_number, greater_than: 0, only_integer: true
+  validates_uniqueness_of :place_number, scope: :movie_id
+  #-----------------------Class Methods-------------------------
+  def self.create_tickets_for_movie(movie, price = 50, amount = 60)
+    amount.times{|i| movie.tickets.create(owner: movie.user, price: price, place_number: i + 1) }
+  end
+
+  def self.available_tickets(movie_id)
+    joins('LEFT JOIN order_items on (tickets.id = order_items.ticket_id)').where('order_items.ticket_id IS NULL')
```
Comment (maksymst) [inline app/models/ticket.rb]: much easier to have an attributes to indicate bought ticket

File: app/services/purchase.rb [hunk 11]
```diff
@@ -0,0 +1,14 @@
+class Purchase
+  #-----------Class methods------------
+  def self.move_from_cart_to_order(current_user)
```
Comment (maksymst) [inline app/services/purchase.rb]: move to model

File: app/views/users/show.html.haml [hunk 12]
```diff
@@ -19,3 +19,19 @@
       %strong address:
       = @user.address.address
 = link_to 'Go to All Users', users_path,  class: 'btn btn-primary'
+.tickets.row.col-xs-12
```
Comment (maksymst) [inline app/views/users/show.html.haml]: use table tag

File: app/views/users/show.html.haml [hunk 13]
```diff
@@ -19,3 +19,19 @@
...
+  - @order_items.each do |order_item|
+    .panel.panel-default
+      .panel-body
+        %p
+          Movie name:
+          = order_item.ticket.name
+
+        %p
+          Price:
+          = number_to_currency order_item.ticket.price
```
Comment (maksymst) [inline app/views/users/show.html.haml]: use `order_item.price`

File: app/views/users/show.html.haml [hunk 14]
```diff
@@ -19,3 +19,19 @@
...
       = @user.address.address
 = link_to 'Go to All Users', users_path,  class: 'btn btn-primary'
+.tickets.row.col-xs-12
+  %h1 Tickets
+  - @order_items.each do |order_item|
+    .panel.panel-default
+      .panel-body
+        %p
+          Movie name:
+          = order_item.ticket.name
```
Comment (maksymst) [inline app/views/users/show.html.haml]: use delegation

File: app/views/users/show.html.haml [hunk 15]
```diff
@@ -19,3 +19,19 @@
...
+
+        %p
+          Price:
+          = number_to_currency order_item.ticket.price
+        %p
+          Place number:
+          = order_item.ticket.place_number
```
Comment (maksymst) [inline app/views/users/show.html.haml]: use delegation

File: db/migrate/20150604123231_create_tickets.rb [hunk 16]
```diff
@@ -0,0 +1,11 @@
+class CreateTickets < ActiveRecord::Migration
+  def change
+    create_table :tickets do |t|
+      t.decimal :price, precision: 8, scale: 2
+      t.belongs_to :movie, index: true
+      t.belongs_to :owner, index: true
```
Comment (maksymst) [inline db/migrate/20150604123231_create_tickets.rb]: movie has an owner

File: spec/models/ticket_spec.rb [hunk 17]
```diff
@@ -0,0 +1,45 @@
+require 'rails_helper'
+
+RSpec.describe Ticket, type: :model do
+  describe 'association' do
+    it { is_expected.to belong_to(:movie) }
+    it { is_expected.to belong_to(:owner) }
+    it { is_expected.to have_many(:cart_items) }
+    it { is_expected.to have_many(:order_items) }
+  end
```
Comment (maksymst) [inline spec/models/ticket_spec.rb:8]: add new line below

File: spec/models/ticket_spec.rb [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+  describe 'validations' do
+    it 'has valid factory' do
+      expect(build(:ticket)).to be_valid
+    end
```
Comment (maksymst) [inline spec/models/ticket_spec.rb:13]: add new line below

File: spec/models/ticket_spec.rb [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+    it { is_expected.to validate_presence_of(:price) }
+    it { is_expected.to validate_presence_of(:place_number) }
+    it { is_expected.to validate_numericality_of(:place_number).is_greater_than(0).only_integer }
+    it { is_expected.to validate_numericality_of(:price).is_greater_than(0) }
+    it { is_expected.to validate_uniqueness_of(:place_number) }
+  end
+
+  describe '.create_tickets_for_movie' do
+    let(:movie) { create :movie }
+    let(:ticket) { Ticket }
```
Comment (MaximSt) [inline spec/models/ticket_spec.rb]: add new line below

File: spec/models/ticket_spec.rb [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+    it 'creates 60 tickets for movie' do
+      expect{ ticket.create_tickets_for_movie(movie) }.to change{ ticket.count }.by(60)
+    end
```
Comment (maksymst) [inline spec/models/ticket_spec.rb]: add new line below

File: spec/models/ticket_spec.rb [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+    it 'creates 5 tickets for movie' do
+      expect{ ticket.create_tickets_for_movie(movie, 30, 5) }.to change{ ticket.count }.by(5)
+    end
+  end
+
+  describe '.available_tickets' do
+    let(:movie) { create :movie }
+    before do
```
Comment (maksymst) [inline spec/models/ticket_spec.rb:28]: add new line below

File: spec/models/ticket_spec.rb [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+      Ticket.create_tickets_for_movie(movie, 20, 5)
+      create :order_item, ticket: first_ticket
+    end
```
Comment (maksymst) [inline spec/models/ticket_spec.rb:33]: add new line below

File: spec/models/ticket_spec.rb [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+    let(:first_ticket) { Ticket.first }
```
Comment (maksymst) [inline spec/models/ticket_spec.rb]: add new line below

File: spec/models/ticket_spec.rb [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+    it "hasn't first ticket in available tickets." do
```
Comment (maksymst) [inline spec/models/ticket_spec.rb]: it does not have

File: spec/services/purchase_spec.rb [hunk 25]
```diff
@@ -0,0 +1,14 @@
+require 'rails_helper'
+RSpec.describe Purchase do
+  describe '.move_from_cart_to_order' do
+    let(:cart) { create :cart }
+    let!(:user) { create :user, cart: cart }
+    let!(:admin) { create(:user) }
+    before { create :cart_item, cart: cart }
+
+    it 'moves items from cart to order' do
```
Comment (maksymst) [inline spec/services/purchase_spec.rb]: it also takes off credits

File: app/controllers/orders_controller.rb [hunk 26]
```diff
@@ -0,0 +1,11 @@
+class OrdersController < ApplicationController
+  def create
+    if current_user.can_afford?
+      current_user.move_from_cart_to_order
```
Comment (maksymst) [inline app/controllers/orders_controller.rb]: `purchase_cart` is much simpler than `move_from_cart_to_order`

File: app/models/cart.rb [hunk 27]
```diff
@@ -0,0 +1,14 @@
+class Cart < ActiveRecord::Base
+  #-----------------------Associations--------------------------
+  has_many :cart_items, dependent: :destroy
+  belongs_to :user
+
+  #-----------------------Instance methods----------------------
+  def quantity
+    cart_items.inject(0){|result, current| result + current.quantity}
```
Comment (maksymst) [inline app/models/cart.rb]: use sql

File: app/models/cart.rb [hunk 28]
```diff
@@ -0,0 +1,14 @@
...
+  end
+
+  def price
+    cart_items.inject(0){|result, current| result + current.ticket.price}
```
Comment (maksymst) [inline app/models/cart.rb]: use sql

File: app/models/movie.rb [hunk 29]
```diff
@@ -6,7 +6,10 @@ class Movie < ActiveRecord::Base
   has_many :uploads, dependent: :destroy
   has_many :daily_statistics, dependent: :destroy
   has_many :monthly_statistics, dependent: :destroy
-
+  has_many :tickets, dependent: :destroy
+  attr_writer :ticket_callback
+  #------------------------------Callbacks---------------------------------
+  after_create { Ticket.create_tickets_for_movie(self) unless @ticket_callback }
```
Comment (maksymst) [inline app/models/movie.rb]: refactor

File: app/models/ticket.rb [hunk 30]
```diff
@@ -0,0 +1,22 @@
...
+  has_many :order_items, dependent: :destroy
+
+  #-----------------------Validations---------------------------
+  validates_presence_of :movie, :price, :place_number
+  validates_numericality_of :price, greater_than: 0
+  validates_numericality_of :place_number, greater_than: 0, only_integer: true
+  validates_uniqueness_of :place_number, scope: :movie_id
+  #-----------------------Class Methods-------------------------
+  def self.create_tickets_for_movie(movie, price = 50, amount = 60)
+    amount.times{|i| movie.tickets.create(price: price, place_number: i + 1) }
```
Comment (maksymst) [inline app/models/ticket.rb]: refactor to use one query

File: app/models/transaction.rb [hunk 31]
```diff
@@ -3,6 +3,7 @@ class Transaction < ActiveRecord::Base
   belongs_to :creditor, foreign_key: :creditor_id, class_name: 'User'
   belongs_to :debitor, foreign_key: :debitor_id, class_name: 'User'
   belongs_to :user
+  has_one :order_item
```
Comment (maksymst) [inline app/models/transaction.rb]: dependent is missing

File: app/models/user.rb [hunk 32]
```diff
@@ -18,6 +18,9 @@ class User < ActiveRecord::Base
   has_many :debit_transactions, foreign_key: :debitor_id, class_name: 'Transaction', dependent: :destroy
   has_many :credit_transactions, foreign_key: :creditor_id, class_name: 'Transaction', dependent: :destroy
   has_many :admin_transactions, foreign_key: :user_id, class_name: 'Transaction', dependent: :destroy
+  has_many :orders
```
Comment (maksymst) [inline app/models/user.rb]: dependent is missing

File: app/models/user.rb [hunk 33]
```diff
@@ -47,6 +50,37 @@ def transfer_money(obtainer_id, amount)
...
     admin_transactions.create(debitor_id: debitor_id, amount: amount, memo: memo)
   end
+
+  def can_afford?
+    balance - cart.cart_items.joins(:ticket).sum(:price) >= 0
+  end
+
+  def move_from_cart_to_order
+    order = orders.create
+    User.transaction do
```
Comment (maksymst) [inline app/models/user.rb]: `transaction do`

File: app/models/user.rb [hunk 34]
```diff
@@ -47,6 +50,37 @@ def transfer_money(obtainer_id, amount)
...
+      cart.cart_items.find_each do |item|
+        transaction = transfer_money(item.ticket.movie.user_id, item.ticket.price)
+        order.order_items.create(order: order, price: item.ticket.price, ticket: item.ticket, transaction_id: transaction.id)
+        item.destroy
+      end
+    end
+  end
+
+  def add_to_cart(params)
+    if cart
```
Comment (maksymst) [inline app/models/user.rb]: remove condition

File: app/models/user.rb [hunk 35]
```diff
@@ -47,6 +50,37 @@ def transfer_money(obtainer_id, amount)
...
+      add_cart_items_to_cart(params)
+    else
+      create_cart
+      add_cart_items_to_cart(params)
+    end
+  end
+
+  def add_cart_items_to_cart(params)
```
Comment (maksymst) [inline app/models/user.rb]: `add_to_cart` is much simpler

File: app/models/ticket.rb [hunk 36]
```diff
@@ -0,0 +1,22 @@
...
+  end
+
+  def self.available_tickets(movie_id)
```
Comment (maksymst) [inline app/models/ticket.rb]: move to Movie, use specific field to find available tickets

File: app/models/user.rb [hunk 37]
```diff
@@ -47,6 +50,37 @@ def transfer_money(obtainer_id, amount)
...
+    movie_id = params[:id]
+    ticket = Ticket.available_tickets(movie_id).sample
+    quantity = params[:quantity] || 1
+    cart.cart_items.create(ticket: ticket, quantity: quantity)
+  end
```
Comment (maksymst) [inline app/models/user.rb]: add new line

File: app/views/carts/show.haml [hunk 38]
```diff
@@ -0,0 +1,24 @@
...
+    %h1 Cart
+    %table.table.table-hover
+      %tr.info
+        %th Item
+        %th Quantity
+        %th Price for one item
+        %th Sum
+      - @cart.cart_items.each do |item|
+        %tr{ id: "item-#{item.id}" }
+          %td= item.ticket.movie.name
```
Comment (maksymst) [inline app/views/carts/show.haml]: use delegator

File: app/views/carts/show.haml [hunk 39]
```diff
@@ -0,0 +1,24 @@
...
+          %td= item.quantity
+          %td= number_to_currency(item.ticket.price)
```
Comment (maksymst) [inline app/views/carts/show.haml:14]: use delegator

File: app/models/cart_item.rb [hunk 40]
```diff
@@ -0,0 +1,12 @@
+class CartItem < ActiveRecord::Base
+  #-----------------------Associations--------------------------
+  belongs_to :cart
+  belongs_to :ticket
+  #-----------------------Validations---------------------------
+  validates_presence_of :cart, :ticket
+  validates_numericality_of :quantity, greater_than_or_equal_to: 1, only_integer: true
+  #-------------------------Instance methods--------------------
+  def calculate_sum
```
Comment (maksymst) [inline app/models/cart_item.rb]: `total` is much simpler

File: app/views/carts/show.haml [hunk 41]
```diff
@@ -0,0 +1,24 @@
...
+          %td
+            = number_to_currency(item.calculate_sum)
+            = link_to 'Remove', cart_item_path(item), method: :delete, data: { confirm: 'Are you sure?' }, class: 'btn btn-danger'
+        %tr
+          %td
+            = link_to 'Buy', orders_path, method: :post, class: 'btn btn-info'
```
Comment (maksymst) [inline app/views/carts/show.haml]: Must be a button

File: app/models/movie.rb [hunk 42]
```diff
@@ -6,7 +6,10 @@ class Movie < ActiveRecord::Base
   has_many :uploads, dependent: :destroy
   has_many :daily_statistics, dependent: :destroy
   has_many :monthly_statistics, dependent: :destroy
-
+  has_many :tickets, dependent: :destroy
+  attr_writer :ticket_callback
```
Comment (maksymst) [inline app/models/movie.rb]: rename

File: spec/models/cart_item_spec.rb [hunk 43]
```diff
@@ -0,0 +1,17 @@
...
+RSpec.describe CartItem, type: :model do
+  describe 'associations' do
+    it { is_expected.to belong_to(:cart)}
+    it { is_expected.to belong_to(:ticket)}
+  end
+
+  describe 'validations' do
+    it 'has valid factory' do
+      expect(build(:cart_item)).to be_valid
+    end
```
Comment (MaximSt) [inline spec/models/cart_item_spec.rb]: add new line below

File: spec/models/order_item_spec.rb [hunk 44]
```diff
@@ -0,0 +1,19 @@
...
+  describe 'associations' do
+    it { is_expected.to belong_to(:order)}
+    it { is_expected.to belong_to(:ticket)}
+    it { is_expected.to belong_to(:order_transaction) }
+  end
+
+  describe 'validations' do
+    it 'has valid factory' do
+      expect(build :order_item).to be_valid
+    end
```
Comment (MaximSt) [inline spec/models/order_item_spec.rb]: add new line below

File: spec/models/order_spec.rb [hunk 45]
```diff
@@ -0,0 +1,15 @@
...
+RSpec.describe Order, type: :model do
+  describe 'associations' do
+    it { is_expected.to belong_to(:user) }
+    it { is_expected.to have_many(:order_items) }
+  end
+
+  describe 'validations' do
+    it 'has valid factory' do
+      expect(build :order).to be_valid
+    end
```
Comment (MaximSt) [inline spec/models/order_spec.rb]: add new line below

File: app/models/movie.rb [hunk 46]
```diff
@@ -14,6 +19,18 @@ class Movie < ActiveRecord::Base
 
   paginates_per 10
 
+  #----------------------------Instance methods----------------------------
+  def create_tickets(price = 50, amount = 60)
```
Comment (MaximSt) [inline app/models/movie.rb]:
```ruby
def create_tickets(price = 50, amount = 60)
  tickets.create((1..amount).inject([]){|res, i| p i; res << {place_number: i}}) do |t|
    t.price = price
  end
end
```

File: app/models/movie.rb [hunk 47]
```diff
@@ -14,6 +19,18 @@ class Movie < ActiveRecord::Base
...
+    tickets_params = []
+    amount.times{ |i| tickets_params << { price: price, place_number: i + 1 } }
+    tickets.create(tickets_params)
+  end
+
+  def available_tickets
+    tickets.where(movie_id: id, bought: false)
```
Comment (maksymst) [inline app/models/movie.rb]: wtf?

File: app/models/order_item.rb [hunk 48]
```diff
@@ -0,0 +1,11 @@
+class OrderItem < ActiveRecord::Base
+  #-----------------------Associations--------------------------
+  belongs_to :order
+  belongs_to :ticket
+  belongs_to :order_transaction, foreign_key: :transaction_id, class_name: 'Transaction'
+  delegate :name, to: :ticket, allow_nil: true
+  delegate :place_number, to: :ticket
```
Comment (maksymst) [inline app/models/order_item.rb]: move to prev line

File: app/models/cart_item.rb [hunk 49]
```diff
@@ -0,0 +1,15 @@
+class CartItem < ActiveRecord::Base
+  #-----------------------Associations--------------------------
+  belongs_to :cart
+  belongs_to :ticket
+
+  delegate :name, to: :ticket
+  delegate :price, to: :ticket
```
Comment (maksymst) [inline app/models/cart_item.rb]: move to prev line

File: app/models/user.rb [hunk 50]
```diff
@@ -47,6 +50,30 @@ def transfer_money(obtainer_id, amount)
   def add_money(debitor_id, amount, memo)
     admin_transactions.create(debitor_id: debitor_id, amount: amount, memo: memo)
   end
+
+  def can_afford?
+    balance - cart.cart_items.joins(:ticket).sum(:price) >= 0
+  end
+
+  def purchase_cart
```
Comment (maksymst) [inline app/models/user.rb:59]: `item.ticket` used 5 times

File: app/models/user.rb [hunk 51]
```diff
@@ -47,6 +50,30 @@ def transfer_money(obtainer_id, amount)
...
+        item.ticket.update_attribute(:bought, true)
+        transaction = transfer_money(item.ticket.movie.user_id, item.ticket.price)
+        order.order_items.create(order: order, price: item.ticket.price, ticket: item.ticket, transaction_id: transaction.id)
+        item.destroy
+      end
+    end
+  end
+
+  def add_to_cart(params, movie)
+    cart = create_cart unless cart
```
Comment (maksymst) [inline app/models/user.rb]: `cart ||= create_cart`

File: app/views/users/show.html.haml [hunk 52]
```diff
@@ -1,21 +1,42 @@
...
+        %strong Email:
+        = @user.email
+      .form-group
+        %strong City:
+        = @user.address.city
+      .form-group
+        %strong Postal Code:
+        = @user.address.postal_code
+      .form-group
+        %strong address:e
```
Comment (maksymst) [inline app/views/users/show.html.haml]: typo

File: spec/models/movie_spec.rb [hunk 53]
```diff
@@ -5,5 +5,18 @@
...
+  end
+
+  describe '.create_tickets' do
+    let(:movie) { create :movie }
+
+    it 'creates 60 tickets for movie' do
+      expect{ movie.create_tickets }.to change{ Ticket.count }.by(60)
+    end
+
+    it 'creates 5 tickets for movie' do
```
Comment (maksymst) [inline spec/models/movie_spec.rb]: what the purpose of this test?

File: app/models/movie.rb [hunk 54]
```diff
@@ -14,6 +19,18 @@ class Movie < ActiveRecord::Base
...
 
+  #----------------------------Instance methods----------------------------
+  def create_tickets(price = 50, amount = 60)
+    tickets.create((1..amount).inject([]){|res, i| res << {place_number: i}}) do |t|
+      t.price = price
+    end
+  end
+
+  def available_tickets
+    tickets.where( bought: false, reserved_by_id: nil )
```
Comment (maksymst) [inline app/models/movie.rb]: move to Ticket as a scope. remove this method

File: app/models/user.rb [hunk 55]
```diff
@@ -47,6 +51,37 @@ def transfer_money(obtainer_id, amount)
   def add_money(debitor_id, amount, memo)
     admin_transactions.create(debitor_id: debitor_id, amount: amount, memo: memo)
   end
+
+  def can_afford?
+    balance - cart.cart_items.joins(:ticket).sum(:price) >= 0
```
Comment (maksymst) [inline app/models/user.rb]: refactor

File: spec/factories/cart_item.rb [hunk 56]
```diff
@@ -0,0 +1,7 @@
+FactoryGirl.define do
+  factory :cart_item, class: 'CartItem' do
+    association :cart
```
Comment (maksymst) [inline spec/factories/cart_item.rb]: just `cart` is fine

File: spec/factories/cart_item.rb [hunk 57]
```diff
@@ -0,0 +1,7 @@
...
+    association :ticket
```
Comment (maksymst) [inline spec/factories/cart_item.rb]: just `ticket` is fine

File: spec/factories/order.rb [hunk 58]
```diff
@@ -0,0 +1,5 @@
+FactoryGirl.define do
+  factory :order, class: 'Order' do
+    association :user
```
Comment (maksymst) [inline spec/factories/order.rb]: just `user` is fine

File: spec/models/ticket_spec.rb [hunk 59]
```diff
@@ -0,0 +1,39 @@
...
+  describe '.available_tickets' do
+    let(:movie) { create :movie }
+    let(:first_ticket) { Ticket.first }
+    let(:user) { create :user }
+
+    before do
+      movie.create_tickets(20, 5)
+      first_ticket.update_attribute(:bought, true)
+      user.reserved_tickets << first_ticket
+      create :order_item, ticket: first_ticket
```
Comment (maksymst) [inline spec/models/ticket_spec.rb:32]: redundant

File: spec/models/ticket_spec.rb [hunk 60]
```diff
@@ -0,0 +1,39 @@
...
+
+  describe '.available_tickets' do
+    let(:movie) { create :movie }
+    let(:first_ticket) { Ticket.first }
+    let(:user) { create :user }
+
+    before do
+      movie.create_tickets(20, 5)
+      first_ticket.update_attribute(:bought, true)
+      user.reserved_tickets << first_ticket
```
Comment (maksymst) [inline spec/models/ticket_spec.rb:31]: redundant

File: spec/models/ticket_spec.rb [hunk 61]
```diff
@@ -0,0 +1,39 @@
...
+  end
+
+  describe '.available_tickets' do
+    let(:movie) { create :movie }
+    let(:first_ticket) { Ticket.first }
+    let(:user) { create :user }
+
+    before do
+      movie.create_tickets(20, 5)
+      first_ticket.update_attribute(:bought, true)
```
Comment (maksymst) [inline spec/models/ticket_spec.rb:30]: redundant

File: spec/models/ticket_spec.rb [hunk 62]
```diff
@@ -0,0 +1,39 @@
...
+    it { is_expected.to validate_uniqueness_of(:place_number) }
+  end
+
+  describe '.available_tickets' do
+    let(:movie) { create :movie }
+    let(:first_ticket) { Ticket.first }
+    let(:user) { create :user }
+
+    before do
+      movie.create_tickets(20, 5)
```
Comment (maksymst) [inline spec/models/ticket_spec.rb:29]: redundant

File: spec/models/ticket_spec.rb [hunk 63]
```diff
@@ -0,0 +1,39 @@
...
+    it { is_expected.to validate_presence_of(:price) }
+    it { is_expected.to validate_presence_of(:place_number) }
+    it { is_expected.to validate_numericality_of(:place_number).is_greater_than(0).only_integer }
+    it { is_expected.to validate_numericality_of(:price).is_greater_than(0) }
+    it { is_expected.to validate_uniqueness_of(:place_number) }
+  end
+
+  describe '.available_tickets' do
+    let(:movie) { create :movie }
+    let(:first_ticket) { Ticket.first }
```
Comment (maksymst) [inline spec/models/ticket_spec.rb:25]: create ticket here with all necessary attributes

**Pull request closed without merging**